### PR TITLE
SLVS-1997 Parse settings.json comma separated sonarlint.analysisExcludesStandalone

### PR DESCRIPTION
**Pull request by vnaskos-sonar**
[SLVS-1997](https://sonarsource.atlassian.net/browse/SLVS-1997)

[SLVS-1997]: https://sonarsource.atlassian.net/browse/SLVS-1997?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Core/Helpers/CommaSeparatedStringArrayConverter.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    public override void WriteJson(JsonWriter writer, string[] value, JsonSerializer serializer) => writer.WriteValue(string.Join(",", value));
+
+    public override string[] ReadJson(
+        JsonReader reader,
+        Type objectType,
+        string[] existingValue,
+        bool hasExistingValue,
+        JsonSerializer serializer)
+    {
+        var value = reader.Value?.ToString();
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Core/Helpers/CommaSeparatedStringArrayConverter.cs:16]: probably should ignore/throw on all non-strings here

File: src/Core/Helpers/CommaSeparatedStringArrayConverter.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+﻿using Newtonsoft.Json;
+
+namespace SonarLint.VisualStudio.Core.Helpers;
+
+public class CommaSeparatedStringArrayConverter : JsonConverter<string[]>
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Core/Helpers/CommaSeparatedStringArrayConverter.cs:5]: We had some problems with the generic base class on older VS versions (they use an outdated newtonsoft json version which does not have it). Run this code on the vs 17.0 to test if it still breaks. If it does, you will need to change it to the non-generic version

File: src/Core/UserRuleSettings/AnalysisSettings.cs [hunk 3]
```diff
@@ -18,70 +18,67 @@
...
-            "level": "on",
-            "Parameters": {
-              "key1": "value1",
-              "key2": "value2"
-            },
-            "severity": "Critical"
-        },
-    }
-...
+    [JsonProperty("sonarlint.analysisExcludesStandalone", NullValueHandling = NullValueHandling.Ignore)]
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Core/UserRuleSettings/AnalysisSettings.cs:60]: null value ignore, but the default value was not changed from [] to null

File: src/Core.UnitTests/UserRuleSettings/AnalysisSettingsTests.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
+﻿using Newtonsoft.Json;
+using SonarLint.VisualStudio.Core.UserRuleSettings;
+
+namespace SonarLint.VisualStudio.Core.UnitTests.UserRuleSettings;
+
+[TestClass]
+public class AnalysisSettingsTests
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Core.UnitTests/UserRuleSettings/AnalysisSettingsTests.cs:7]: Why not add these tests to `AnalysisSettingsSerializerTests`? Would be nice to have a combined test as well

**Review by georgii-borovinskikh-sonarsource — CHANGES_REQUESTED** [review]

File: src/Core/Helpers/CommaSeparatedStringArrayConverter.cs (hunk 2 above)
Comment (vnaskos-sonar) [reply src/Core/Helpers/CommaSeparatedStringArrayConverter.cs]: Good to know, yup it's still an issue

File: src/Core.UnitTests/UserRuleSettings/AnalysisSettingsTests.cs (hunk 4 above)
Comment (vnaskos-sonar) [reply src/Core.UnitTests/UserRuleSettings/AnalysisSettingsTests.cs:7]: The `AnalysisSettingsSerializerTests` is testing a lot of things at the same time, this is why I created the `AnalysisSettingsTests` to test only for the serialization. I didn't want to refactor that test in this PR, the keep the scope small.

File: src/Core/Helpers/CommaSeparatedStringArrayConverter.cs (hunk 1 above)
Comment (vnaskos-sonar) [reply src/Core/Helpers/CommaSeparatedStringArrayConverter.cs:23]: If it's not a string, it will return null (ignore)

File: src/Core.UnitTests/UserRuleSettings/AnalysisSettingsTests.cs [hunk 5]
```diff
@@ -23,6 +23,22 @@ public void AnalysisSettings_FileExclusions_SerializesCorrectly()
         json.Should().Be(expectedJson);
     }
 
+    [TestMethod]
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Core.UnitTests/UserRuleSettings/AnalysisSettingsTests.cs:26]: still missing tests where both fields are not empty

File: src/Core/Helpers/CommaSeparatedStringArrayConverter.cs (hunk 1 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/Core/Helpers/CommaSeparatedStringArrayConverter.cs:23]: 🤔 not what I would expect

**Review by georgii-borovinskikh-sonarsource — APPROVED** [review]
